Compiling additional dimensions

Hi,
Great work, I wanted to compile dimensions 8 but failed.
Compiling without modification works fine, however adding new template dimensions to the pybind11 module in interface.cpp gives me errors.
On windows I get this error:
```
 1>Checking Build System
        Creating directories for 'eigen'
        Building Custom Rule C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/CMakeLists.txt
        Performing download step (download, verify and extract) for 'eigen'
        -- File already exists but no hash specified (use URL_HASH):
          file='C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/build/temp.win-amd64-cpython-311/Release/eigen-prefix/src/eigen-3.4.0.tar.bz2'
        Old file will be removed and new file downloaded from URL.
        -- Downloading...
           dst='C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/build/temp.win-amd64-cpython-311/Release/eigen-prefix/src/eigen-3.4.0.tar.bz2'
           timeout='none'
           inactivity timeout='none'
        -- Using src='https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.tar.bz2'
        -- Downloading... done
        -- extracting...
             src='C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/build/temp.win-amd64-cpython-311/Release/eigen-prefix/src/eigen-3.4.0.tar.bz2'
             dst='C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/eigenlib'
        -- extracting... [tar xfz]
        -- extracting... [analysis]
        -- extracting... [rename]
        -- extracting... [clean up]
        -- extracting... done
        No update step for 'eigen'
        No patch step for 'eigen'
        No configure step for 'eigen'
        No build step for 'eigen'
        No install step for 'eigen'
        Completed 'eigen'
        Building Custom Rule C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/CMakeLists.txt
        test_kdtree.vcxproj -> C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\Release\test_kdtree.exe
        Building Custom Rule C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/CMakeLists.txt
           Creating library C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/build/temp.win-amd64-cpython-311/Release/Release/torch_knn.lib and object C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/torch_kdtree/build/temp.win-amd64-cpython-311/Release/Release/torch_knn.exp
      interface.obj : error LNK2019: unresolved external symbol "public: __cdecl PartitionInfo<double,8,1>::PartitionInfo<double,8,1>(class std::vector<struct Partition<double>,class std::allocator<struct Partition<double> > > &&,class std::vector<struct PartitionLeaf<double,8>,class std::allocator<struct PartitionLeaf<double,8> > > &&,int *,int)" (??0?$PartitionInfo@N$07$00@@QEAA@$$QEAV?$vector@U?$Partition@N@@V?$allocator@U?$Partition@N@@@std@@@std@@$$QEAV?$vector@U?$PartitionLeaf@N$07@@V?$allocator@U?$PartitionLeaf@N$07@@@std@@@2@PEAHH@Z) referenced in function "struct PartitionInfo<double,8,1> __cdecl createKDTree<double,8>(double const *,int,int)" (??$createKDTree@N$07@@YA?AU?$PartitionInfo@N$07$00@@PEBNHH@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "public: __cdecl PartitionInfo<float,8,1>::PartitionInfo<float,8,1>(class std::vector<struct Partition<float>,class std::allocator<struct Partition<float> > > &&,class std::vector<struct PartitionLeaf<float,8>,class std::allocator<struct PartitionLeaf<float,8> > > &&,int *,int)" (??0?$PartitionInfo@M$07$00@@QEAA@$$QEAV?$vector@U?$Partition@M@@V?$allocator@U?$Partition@M@@@std@@@std@@$$QEAV?$vector@U?$PartitionLeaf@M$07@@V?$allocator@U?$PartitionLeaf@M$07@@@std@@@2@PEAHH@Z) referenced in function "struct PartitionInfo<float,8,1> __cdecl createKDTree<float,8>(float const *,int,int)" (??$createKDTree@M$07@@YA?AU?$PartitionInfo@M$07$00@@PEBMHH@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "void __cdecl KDTreeKNNGPUSearch<double,double,8>(struct PartitionInfo<double,8,0> *,int,class std::array<double,8> const *,double *,int *,int)" (??$KDTreeKNNGPUSearch@NN$07@@YAXPEAU?$PartitionInfo@N$07$0A@@@HPEBV?$array@N$07@std@@PEANPEAHH@Z) referenced in function "public: void __cdecl KDTree<double,8,1>::query(unsigned __int64,unsigned __int64,int,unsigned __int64,unsigned __int64)" (?query@?$KDTree@N$07$00@@QEAAX_K0H00@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "void __cdecl KDTreeKNNSearch<double,double,8>(struct PartitionInfo<double,8,1> &,int,class std::array<double,8> const *,double *,int *,int)" (??$KDTreeKNNSearch@NN$07@@YAXAEAU?$PartitionInfo@N$07$00@@HPEBV?$array@N$07@std@@PEANPEAHH@Z) referenced in function "public: void __cdecl KDTree<double,8,0>::query(unsigned __int64,unsigned __int64,int,unsigned __int64,unsigned __int64)" (?query@?$KDTree@N$07$0A@@@QEAAX_K0H00@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "void __cdecl KDTreeKNNGPUSearch<float,float,8>(struct PartitionInfo<float,8,0> *,int,class std::array<float,8> const *,float *,int *,int)" (??$KDTreeKNNGPUSearch@MM$07@@YAXPEAU?$PartitionInfo@M$07$0A@@@HPEBV?$array@M$07@std@@PEAMPEAHH@Z) referenced in function "public: void __cdecl KDTree<float,8,1>::query(unsigned __int64,unsigned __int64,int,unsigned __int64,unsigned __int64)" (?query@?$KDTree@M$07$00@@QEAAX_K0H00@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "void __cdecl KDTreeKNNSearch<float,float,8>(struct PartitionInfo<float,8,1> &,int,class std::array<float,8> const *,float *,int *,int)" (??$KDTreeKNNSearch@MM$07@@YAXAEAU?$PartitionInfo@M$07$00@@HPEBV?$array@M$07@std@@PEAMPEAHH@Z) referenced in function "public: void __cdecl KDTree<float,8,0>::query(unsigned __int64,unsigned __int64,int,unsigned __int64,unsigned __int64)" (?query@?$KDTree@M$07$0A@@@QEAAX_K0H00@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "void __cdecl freePartitionFromGPU<double,8>(struct PartitionInfo<double,8,0> *)" (??$freePartitionFromGPU@N$07@@YAXPEAU?$PartitionInfo@N$07$0A@@@@Z) referenced in function "public: __cdecl std::_Temporary_owner<struct KDTree<double,8,1> >::~_Temporary_owner<struct KDTree<double,8,1> >(void)" (??1?$_Temporary_owner@U?$KDTree@N$07$00@@@std@@QEAA@XZ) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "void __cdecl freePartitionFromGPU<float,8>(struct PartitionInfo<float,8,0> *)" (??$freePartitionFromGPU@M$07@@YAXPEAU?$PartitionInfo@M$07$0A@@@@Z) referenced in function "public: __cdecl std::_Temporary_owner<struct KDTree<float,8,1> >::~_Temporary_owner<struct KDTree<float,8,1> >(void)" (??1?$_Temporary_owner@U?$KDTree@M$07$00@@@std@@QEAA@XZ) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "struct PartitionInfo<double,8,0> * __cdecl copyPartitionToGPU<double,8>(struct PartitionInfo<double,8,1> const &)" (??$copyPartitionToGPU@N$07@@YAPEAU?$PartitionInfo@N$07$0A@@@AEBU?$PartitionInfo@N$07$00@@@Z) referenced in function "struct KDTree<double,8,1> * __cdecl pybind11::detail::initimpl::construct_or_initialize<struct KDTree<double,8,1>,class pybind11::array_t<double,16>,int,0>(class pybind11::array_t<double,16> &&,int &&)" (??$construct_or_initialize@U?$KDTree@N$07$00@@V?$array_t@N$0BA@@pybind11@@H$0A@@initimpl@detail@pybind11@@YAPEAU?$KDTree@N$07$00@@$$QEAV?$array_t@N$0BA@@2@$$QEAH@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      interface.obj : error LNK2019: unresolved external symbol "struct PartitionInfo<float,8,0> * __cdecl copyPartitionToGPU<float,8>(struct PartitionInfo<float,8,1> const &)" (??$copyPartitionToGPU@M$07@@YAPEAU?$PartitionInfo@M$07$0A@@@AEBU?$PartitionInfo@M$07$00@@@Z) referenced in function "struct KDTree<float,8,1> * __cdecl pybind11::detail::initimpl::construct_or_initialize<struct KDTree<float,8,1>,class pybind11::array_t<float,16>,int,0>(class pybind11::array_t<float,16> &&,int &&)" (??$construct_or_initialize@U?$KDTree@M$07$00@@V?$array_t@M$0BA@@pybind11@@H$0A@@initimpl@detail@pybind11@@YAPEAU?$KDTree@M$07$00@@$$QEAV?$array_t@M$0BA@@2@$$QEAH@Z) [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\Release\torch_knn.cp311-win_amd64.pyd : fatal error LNK1120: 10 unresolved externals [C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\temp.win-amd64-cpython-311\Release\torch_knn.vcxproj]
      Building extension...
      Building extension #2...
      Build temp dir:  build\temp.win-amd64-cpython-311\Release C:\OneDrive - UvA\Jaar 5\Thesis\Side-chain auto-encoder\torch_kdtree\build\lib.win-amd64-cpython-311\
      ['cmake', 'C:\\OneDrive - UvA\\Jaar 5\\Thesis\\Side-chain auto-encoder\\torch_kdtree', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\\OneDrive - UvA\\Jaar 5\\Thesis\\Side-chain auto-encoder\\torch_kdtree\\build\\lib.win-amd64-cpython-311', '-DPYTHON_EXECUTABLE=C:\\Users\\R35re\\anaconda3\\envs\\sidechain\\python.exe', '-DEXAMPLE_VERSION_INFO=1.0', '-DCMAKE_BUILD_TYPE=Release', '-DPIP_BUILD=ON', '-A', 'x64']
      ['cmake', '--build', '.', '--config', 'Release']
      Traceback (most recent call last):
        File "C:\Users\R35re\anaconda3\envs\sidechain\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 389, in <module>
          main()
        File "C:\Users\R35re\anaconda3\envs\sidechain\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 373, in main
          json_out["return_val"] = hook(**hook_input["kwargs"])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\R35re\anaconda3\envs\sidechain\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 280, in build_wheel
          return _build_backend().build_wheel(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\build_meta.py", line 438, in build_wheel
          return _build(['bdist_wheel', '--dist-info-dir', str(metadata_directory)])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\build_meta.py", line 426, in _build
          return self._build_with_temp_dir(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\build_meta.py", line 407, in _build_with_temp_dir
          self.run_setup()
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\build_meta.py", line 320, in run_setup
          exec(code, locals())
        File "<string>", line 145, in <module>
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\__init__.py", line 117, in setup
          return distutils.core.setup(**attrs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 186, in setup
          return run_commands(dist)
                 ^^^^^^^^^^^^^^^^^^
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 202, in run_commands
          dist.run_commands()
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 983, in run_commands
          self.run_command(cmd)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\dist.py", line 999, in run_command
          super().run_command(command)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 1002, in run_command
          cmd_obj.run()
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\command\bdist_wheel.py", line 379, in run
          self.run_command("build")
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 339, in run_command
          self.distribution.run_command(command)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\dist.py", line 999, in run_command
          super().run_command(command)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 1002, in run_command
          cmd_obj.run()
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\command\build.py", line 136, in run
          self.run_command(cmd_name)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 339, in run_command
          self.distribution.run_command(command)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\dist.py", line 999, in run_command
          super().run_command(command)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 1002, in run_command
          cmd_obj.run()
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\command\build_ext.py", line 99, in run
          _build_ext.run(self)
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\command\build_ext.py", line 365, in run
          self.build_extensions()
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\command\build_ext.py", line 481, in build_extensions
          self._build_extensions_serial()
        File "C:\Users\R35re\AppData\Local\Temp\pip-build-env-8zd0889u\overlay\Lib\site-packages\setuptools\_distutils\command\build_ext.py", line 507, in _build_extensions_serial
          self.build_extension(ext)
        File "<string>", line 134, in build_extension
        File "C:\Users\R35re\anaconda3\envs\sidechain\Lib\subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release']' returned non-zero exit status 1.
```
On linux it compiles fine but when trying to import I get this error:
```Traceback (most recent call last):
  File "/gpfs/home2/rdenbraber/Sidechain-auto-encoder/test.py", line 1, in <module>
    import torch_kdtree
  File "/gpfs/home2/rdenbraber/Sidechain-auto-encoder/sidechain/lib/python3.11/site-packages/torch_kdtree/__init__.py", line 1, in <module>
    from .nn_distance import build_kd_tree, gpu_available
  File "/gpfs/home2/rdenbraber/Sidechain-auto-encoder/sidechain/lib/python3.11/site-packages/torch_kdtree/nn_distance.py", line 16, in <module>
    raise(err)
  File "/gpfs/home2/rdenbraber/Sidechain-auto-encoder/sidechain/lib/python3.11/site-packages/torch_kdtree/nn_distance.py", line 6, in <module>
    from . import torch_knn
ImportError: /gpfs/home2/rdenbraber/Sidechain-auto-encoder/sidechain/lib/python3.11/site-packages/torch_kdtree/torch_knn.cpython-311-x86_64-linux-gnu.so: undefined symbol: _Z18KDTreeKNNGPUSearchIddLh8EEvP13PartitionInfoIT_XT1_ELb0EEiPKSt5arrayIS1_XT1_EEPS1_Pii
```

Hopefully there is some fix for this as having additional dimensions would be great

There are two error message that I found:
>       CMake Error at CMakeLists.txt:5 (add_subdirectory):
>        The source directory
>
>         C:/OneDrive - UvA/Jaar 5/Thesis/Side-chain auto-encoder/kd_tree_test/torch_kdtree/pybind11
>
>        does not contain a CMakeLists.txt file.
and
>      CMake Error at CMakeLists.txt:130 (set_target_properties):
>        set_target_properties called with incorrect number of arguments.

Regarding the first problem, did you run
```bash
git submodule init
git submodule update
```
as written in README.md? It seems pybind11 directory is empty.
The second one puzzles me a bit. It works on my machine with CMake 3.21.

--- a/README.md
+++ b/README.md
@@ -56,7 +56,7 @@ Prerequisites
 - Numpy (installed with `setuptools`)
 - Torch (installed with `setuptools`)
 - Cuda
-- g++, or Visual Studio (MacOSX is untested)
+- g++, or Visual Studio x64 (MacOSX is untested)
 - CMake
 
 Build Instruction
@@ -92,7 +92,7 @@ To run the benchmark on your computer, simply run `python benchmark/benchmark.py
 The dimension of the KD-Tree are compile time dynamic, meaning that the dimensions to be queried need to be known at compile time. By default, the library is compiled for d in [1, 2, 3]. You can add additional dimensions by adding new template dimensions in three places.
 
 To add dimensionality 8 for example, you have to add the following code snippets  
-`interface.cpp` (line 115)
+`src/interface.cpp` (line 115)
 ```cpp
 KDTREE_INSTANTIATION(float, 8, false, "KDTreeCPU8DF");
 KDTREE_INSTANTIATION(double, 8, false, "KDTreeCPU8D");
@@ -100,7 +100,7 @@ KDTREE_INSTANTIATION(float, 8, true, "KDTreeGPU8DF");
 KDTREE_INSTANTIATION(double, 8, true, "KDTreeGPU8D");
 ```
 
-In `kdtree_g.cu` (line 476) and `kdtree.cpp` (line 221), you add the same code:
+In `src/kdtree_g.cu` (line 476) and `src/kdtree.cpp` (line 221), you add the same code:
 ```cpp
 KDTREE_INSTANTIATION(float, 8);
 KDTREE_INSTANTIATION(double, 8);

--- a/README.md
+++ b/README.md
@@ -89,14 +89,21 @@ To run the benchmark on your computer, simply run `python benchmark/benchmark.py
 
 # Compiling additional dimensions
 
-The dimension of the KD-Tree are compile time dynamic, meaning that the dimensions to be queried need to be known at compile time. By default, the library is compiled for d in [1, 2, 3]. You can add additional dimensions by adding new template dimensions to the pybind11 module in `interface.cpp` (line 115).
+The dimension of the KD-Tree are compile time dynamic, meaning that the dimensions to be queried need to be known at compile time. By default, the library is compiled for d in [1, 2, 3]. You can add additional dimensions by adding new template dimensions in three places.
 
-To add dimensionality 8 for example, you can add:
+To add dimensionality 8 for example, you have to add the following code snippets  
+`interface.cpp` (line 115)
 ```cpp
-    KDTREE_INSTANTIATION(float, 8, false, "KDTreeCPU8DF");
-    KDTREE_INSTANTIATION(double, 8, false, "KDTreeCPU8D");
-    KDTREE_INSTANTIATION(float, 8, true, "KDTreeGPU8DF");
-    KDTREE_INSTANTIATION(double, 8, true, "KDTreeGPU8D");
+KDTREE_INSTANTIATION(float, 8, false, "KDTreeCPU8DF");
+KDTREE_INSTANTIATION(double, 8, false, "KDTreeCPU8D");
+KDTREE_INSTANTIATION(float, 8, true, "KDTreeGPU8DF");
+KDTREE_INSTANTIATION(double, 8, true, "KDTreeGPU8D");
+```
+
+In `kdtree_g.cu` (line 476) and `kdtree.cpp` (line 221), you add the same code:
+```cpp
+KDTREE_INSTANTIATION(float, 8);
+KDTREE_INSTANTIATION(double, 8);
 ```
 
 This will instantiate the template functions for float and double types both on the CPU and GPU.

--- a/src/kdtree.cpp
+++ b/src/kdtree.cpp
@@ -193,29 +193,6 @@ void KDTreeKNNSearch(PartitionInfo<T, dims>& partition_info,
 	}
 }
 
-template void KDTreeKNNSearch<float, float, 3>(PartitionInfo<float, 3>& partition_info,
-                    const point_i_t nr_query, 
-                    const std::array<float, 3>* points_query, float * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-
-template void KDTreeKNNSearch<double, double, 3>(PartitionInfo<double, 3>& partition_info,
-                    const point_i_t nr_query, 
-                    const std::array<double, 3>* points_query, double * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-
-template void KDTreeKNNSearch<float, float, 2>(PartitionInfo<float, 2>& partition_info,
-                    const point_i_t nr_query, 
-                    const std::array<float, 2>* points_query, float * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-
-template void KDTreeKNNSearch<double, double, 2>(PartitionInfo<double, 2>& partition_info,
-                    const point_i_t nr_query, 
-                    const std::array<double, 2>* points_query, double * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-
-template void KDTreeKNNSearch<float, float, 1>(PartitionInfo<float, 1>& partition_info,
-                    const point_i_t nr_query, 
-                    const std::array<float, 1>* points_query, float * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-
-template void KDTreeKNNSearch<double, double, 1>(PartitionInfo<double, 1>& partition_info,
-                    const point_i_t nr_query, 
-                    const std::array<double, 1>* points_query, double * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
 
 template <typename T, dim_t dims, bool delete_partitions>
 PartitionInfo<T, dims, delete_partitions>::PartitionInfo(std::vector<Partition<T>>&& parts, std::vector<PartitionLeaf<T, dims>>&& leaves_, point_i_t* shuffled_inds_, const point_i_t nr_points_) : 
@@ -228,16 +205,16 @@ nr_partitions(parts.size()), nr_leaves(leaves_.size())
     structured_points = leaves[0].data;
 }
 
-template struct PartitionInfo<float, 1, false>;
-template struct PartitionInfo<float, 2, false>;
-template struct PartitionInfo<float, 3, false>;
-template struct PartitionInfo<double, 1, false>;
-template struct PartitionInfo<double, 2, false>;
-template struct PartitionInfo<double, 3, false>;
 
-template struct PartitionInfo<float, 1, true>;
-template struct PartitionInfo<float, 2, true>;
-template struct PartitionInfo<float, 3, true>;
-template struct PartitionInfo<double, 1, true>;
-template struct PartitionInfo<double, 2, true>;
-template struct PartitionInfo<double, 3, true>;
+#define KDTREE_INSTANTIATION(T, dims) template void KDTreeKNNSearch<T, T, dims>(PartitionInfo<T, dims>& partition_info, \
+                    const point_i_t nr_query,  \
+                    const std::array<T, dims>* points_query, T * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches); \
+                    template struct PartitionInfo<T, dims, false>; \
+                    template struct PartitionInfo<T, dims, true>;
+
+KDTREE_INSTANTIATION(float, 1);
+KDTREE_INSTANTIATION(double, 1);
+KDTREE_INSTANTIATION(float, 2);
+KDTREE_INSTANTIATION(double, 2);
+KDTREE_INSTANTIATION(float, 3);
+KDTREE_INSTANTIATION(double, 3);

--- a/src/kdtree_g.cu
+++ b/src/kdtree_g.cu
@@ -456,53 +456,24 @@ void KDTreeKNNGPUSearch(PartitionInfoDevice<T, dims>* partition_info,
 	#endif*/
 }
 
-template void compQuadrDistLeafPartition<float, float, 3>(const std::array<float, 3>& point, const PartitionLeaf<float, 3>& partition_leaf,
-                                    float* best_dists, point_i_knn_t* best_idx,
-                                    const point_i_knn_t nr_nns_searches);
-                                    
-template void compQuadrDistLeafPartition<double, double, 3>(const std::array<double, 3>& point, const PartitionLeaf<double, 3>& partition_leaf,
-                                    double* best_dists, point_i_knn_t* best_idx,
-                                    const point_i_knn_t nr_nns_searches);
+#define KDTREE_INSTANTIATION(T, dims) template void compQuadrDistLeafPartition<T, T, dims>(const std::array<T, dims>& point, const PartitionLeaf<T, dims>& partition_leaf, \
+				T* best_dists, point_i_knn_t* best_idx, \
+				const point_i_knn_t nr_nns_searches); \
+				template void KDTreeKNNGPUSearch<T, T, dims>(PartitionInfoDevice<T, dims>* partition_info, \
+                    const point_i_t nr_query, \
+                    const std::array<T, dims>* points_query, T * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches); \
+				template PartitionInfoDevice<T, dims>* copyPartitionToGPU(const PartitionInfo<T, dims>& partition_info); \
+				template std::tuple<T*, point_i_t*, T*> copyData<T, dims>(const std::vector<T>& result_dists, const std::vector<point_i_t>& result_idx, const std::vector<std::array<T, dims>>&); \
+				template void freePartitionFromGPU(PartitionInfoDevice<T, dims>* partition_info);
 
-template void KDTreeKNNGPUSearch<float, float, 1>(PartitionInfoDevice<float, 1>* partition_info,
-                    const point_i_t nr_query, 
-                    const std::array<float, 1>* points_query, float * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-template void KDTreeKNNGPUSearch<double, double, 1>(PartitionInfoDevice<double, 1>* partition_info, 
-    const point_i_t nr_query, 
-    const std::array<double, 1>* points_query, double * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-	
-template void KDTreeKNNGPUSearch<float, float, 2>(PartitionInfoDevice<float, 2>* partition_info,
-		const point_i_t nr_query, 
-		const std::array<float, 2>* points_query, float * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-template void KDTreeKNNGPUSearch<double, double, 2>(PartitionInfoDevice<double, 2>* partition_info,
-	const point_i_t nr_query, 
-	const std::array<double, 2>* points_query, double * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
 
-template void KDTreeKNNGPUSearch<float, float, 3>(PartitionInfoDevice<float, 3>* partition_info,
-	const point_i_t nr_query, 
-	const std::array<float, 3>* points_query, float * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-template void KDTreeKNNGPUSearch<double, double, 3>(PartitionInfoDevice<double, 3>* partition_info,
-	const point_i_t nr_query, 
-	const std::array<double, 3>* points_query, double * dist, point_i_t* idx, const point_i_knn_t nr_nns_searches);
-
-template PartitionInfoDevice<float, 1>* copyPartitionToGPU(const PartitionInfo<float, 1>& partition_info);
-template PartitionInfoDevice<float, 2>* copyPartitionToGPU(const PartitionInfo<float, 2>& partition_info);
-template PartitionInfoDevice<float, 3>* copyPartitionToGPU(const PartitionInfo<float, 3>& partition_info);
-template PartitionInfoDevice<double, 1>* copyPartitionToGPU(const PartitionInfo<double, 1>& partition_info);
-template PartitionInfoDevice<double, 2>* copyPartitionToGPU(const PartitionInfo<double, 2>& partition_info);
-template PartitionInfoDevice<double, 3>* copyPartitionToGPU(const PartitionInfo<double, 3>& partition_info);
-
-template std::tuple<float*, point_i_t*, float*> copyData<float, 3>(const std::vector<float>& result_dists, const std::vector<point_i_t>& result_idx, const std::vector<std::array<float, 3>>&);
-template std::tuple<double*, point_i_t*, double*> copyData<double, 3>(const std::vector<double>& result_dists, const std::vector<point_i_t>& result_idx, const std::vector<std::array<double, 3>>&);
-
-template void freePartitionFromGPU(PartitionInfoDevice<float, 1>* partition_info);
-template void freePartitionFromGPU(PartitionInfoDevice<float, 2>* partition_info);
-template void freePartitionFromGPU(PartitionInfoDevice<float, 3>* partition_info);
-template void freePartitionFromGPU(PartitionInfoDevice<double, 1>* partition_info);
-template void freePartitionFromGPU(PartitionInfoDevice<double, 2>* partition_info);
-template void freePartitionFromGPU(PartitionInfoDevice<double, 3>* partition_info);
+KDTREE_INSTANTIATION(float, 1);
+KDTREE_INSTANTIATION(double, 1);
+KDTREE_INSTANTIATION(float, 2);
+KDTREE_INSTANTIATION(double, 2);
+KDTREE_INSTANTIATION(float, 3);
+KDTREE_INSTANTIATION(double, 3);
 
 template void freeGPUArray(float* arr);
 template void freeGPUArray(double* arr);
 template void freeGPUArray(point_i_knn_t* arr);
-

